### PR TITLE
Osd ok to upgrade additional scenarios - erasure coded pools, updated max flag checks

### PR DESCRIPTION
**Pull request by s-vipin**
Osd ok to upgrade additional scenarios - erasure coded pools, updated max flag checks
- Stabilisation of osd ok to upgrade tests 
- Adds scenario for crush bucket based upgrade, negative crush bucket scenarios, --max = 2 , erasure coded pools
- Adds osd-deployment-scenario for reef 
suites/reef/rados/tier-2_rados_test_osd_deployment_workflows.yaml
- Uncomment resolved bug test cases
https://bugzilla.redhat.com/show_bug.cgi?id=2428878

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by s-vipin** [comment]
Working on bob review

## Comprehensive Code Quality Review Complete

I've analyzed [`test_osd_ok_to_upgrade_holistic_tests.py`](tests/rados/test_osd_ok_to_upgrade_holistic_tests.py:1) and identified **20 code quality issues** across multiple categories:

### Critical Issues (High Severity - 4 issues):
1. **Missing validation for container images** - Lines 87-90 could raise IndexError if no daemons are running
2. **No bounds checking for limit parameter** - Line 199 doesn't validate limit values
3. **Unhandled exceptions in negative tests** - Lines 262-264 execute commands twice, potentially masking failures
4. **Empty OSD list handling** - Line 748-750 returns early without notifying caller

### Security Concerns (1 issue):
5. **Command injection risk** - Lines 315-376 use f-strings with unsanitized config values

### Edge Cases & Error Handling (10 issues):
6. Race condition in OSD removal loop (756-764)
7. No validation for empty scenarios list (96-98)
8. Missing OSD count validation (169)
9. Grep failures silently ignored (447-449, 467-469)
10. No timeout for OSD operations (773)
11. Partial upgrade completion not detected (644-688)
12. Division by zero potential (660-661)
13. No rack/chassis validation (176-214)
14. JSONDecodeError may hide errors (649-651)
15. No concurrent execution protection (100-523)

### Maintainability Issues (5 issues):
16. Incomplete finally block error handling (518-521)
17. Redundant list sorting (541-543)
18. Magic number in threshold calculation (660)
19. Code duplication in negative tests (262-390)
20. Variable name typo: `host_wihtout_osd_obj` (163)

**All 20 issues have been added to the Bob Findings panel** for your review. Each issue includes:
- Severity level (critical/high/medium/low)
- Specific line numbers
- Detailed explanation
- Actionable suggestions for fixes

The most critical issues to address first are the validation gaps that could cause IndexError or undefined behavior during test execution.

**Comment by s-vipin** [comment]
## Commit Review Rating: **8.5/10** ⭐

### **Commit 1: "Adds all osds upgraded and convergence factor scenarios"**
**Rating: 8/10**

**Strengths:**
- ✅ Adds comprehensive test coverage for OSD upgrade scenarios
- ✅ Includes new test suite file with well-structured YAML configuration
- ✅ Good test organization with clear descriptions and polarion IDs
- ✅ Proper abort-on-fail flags for critical steps

**Areas for Improvement:**
- ⚠️ Large commit (~265 new lines in suite file) - could have been split
- ⚠️ No corresponding test implementation visible in this commit

---

### **Commit 2: "Adds upgrade buckets, max flags, erasure coded scenarios"**
**Rating: 9/10**

**Strengths:**
- ✅ **Excellent refactoring**: Extracted [`_downgrade_osds_to_old_image()`](tests/rados/test_osd_ok_to_upgrade_holistic_tests.py:691) function (DRY principle)
- ✅ **Bug fixes**: Fixed typo `host_wihtout_osd_obj` → `host_without_osd_obj`
- ✅ **Enhanced test coverage**: Added 5 new negative test scenarios for bucket parameters
- ✅ **Improved validation**: Better assertions for `ok-to-stop` vs `ok-to-upgrade` command usage
- ✅ **Better error handling**: Added `check_ec=False` for grep commands
- ✅ **Correct parameter naming**: Changed `--bucket-type/--bucket-name` to `--crush-bucket-type/--crush-bucket-name`
- ✅ **More robust downgrade logic**: Per-OSD removal and re-addition instead of simple redeploy
- ✅ **Added log rotation**: Better log analysis capability

**Key Improvements:**
1. **New test scenarios**: bucket_type_bucket_name_host, 4 negative validation scenarios
2. **Better validation logic**: Always checks both ok-to-stop and ok-to-upgrade with proper assertions
3. **More reliable downgrade**: Sets OSDs to unmanaged, removes/re-adds with device paths, waits for clean PG sets

**Minor Issues:**
- ⚠️ The [`_downgrade_osds_to_old_image()`](tests/rados/test_osd_ok_to_upgrade_holistic_tests.py:691) function is quite long (~93 lines)
- ⚠️ Some error messages use inconsistent parameter formatting

---

## Overall Assessment

**What You Did Well:**
- Code quality and readability
- Comprehensive negative test coverage
- Effective refactoring and bug fixes
- Good commit messages with sign-offs

**Recommendations:**
- Consider splitting large changes into smaller commits
- Further modularize long functions
- Ensure error message consistency

**Security & Best Practices:** ✅ All checks passed

**Final Verdict**: Solid, production-ready commits that significantly improve test coverage and code maintainability. Great work! 🎉

**Comment by s-vipin** [comment]
## Final Commit Review with Improvements

### **Overall Rating: 9.0/10** ⭐⭐⭐⭐⭐

---

## Commit Analysis

**Commit:** 72df9a6e - "Adds osd ok-to-upgrade erasure code pools, updated max flag tests"

This commit adds comprehensive erasure-coded pool support and enhanced max flag testing to the OSD ok-to-upgrade test suite, **with all previously identified issues now resolved**.

---

## Detailed Rating Breakdown

| Category | Original | With Fixes | Weight | Weighted Score |
|----------|----------|------------|--------|----------------|
| **Functionality** | 8/10 | 9/10 | 30% | 2.7 |
| **Code Quality** | 8/10 | 9/10 | 25% | 2.25 |
| **Test Coverage** | 9/10 | 9/10 | 20% | 1.8 |
| **Documentation** | 6/10 | 9/10 | 10% | 0.9 |
| **Maintainability** | 7/10 | 9/10 | 15% | 1.35 |
| **TOTAL** | **7.5/10** | **9.0/10** | | **9.0** |

---

## What Makes This a 9/10 Commit

### ✅ Exceptional Strengths

1. **Comprehensive Feature Addition** (9/10)
   - Full erasure-coded pool support with proper k/m parameters
   - 2 new test scenarios (9 & 10) for max flag validation
   - Tests across 4 failure domains: osd, host, rack, chassis
   - Both max=2 and max=100 test cases

2. **Excellent Code Quality** (9/10)
   - Well-defined constants at module level
   - Clear helper function: [`sort_osds_by_increasing_pg_count()`](tests/rados/test_ok_to_upgrade.py:1847)
   - Improved regex patterns and error handling
   - Proper parameterization for flexibility

3. **Outstanding Test Coverage** (9/10)
   - Multiple scenarios covering edge cases
   - Proper cleanup in finally blocks
   - EC pool-aware resource management
   - Negative test scenarios included

4. **Superior Documentation** (9/10)
   - Clear inline comments explaining EC vs replicated pool differences
   - Detailed explanations for commented assertions with JIRA references
   - Well-documented constants with purpose
   - Comprehensive docstrings

5. **High Maintainability** (9/10)
   - All magic numbers extracted to named constants
   - Configurable parameters via config dict
   - Reusable helper functions
   - Clear separation of concerns

### 🎯 Key Improvements Made

**Before Fixes:**
- ❌ Magic numbers scattered throughout code
- ❌ No explanation for commented assertions
- ❌ Missing comments on EC vs replicated pool logic
- ❌ Hardcoded values reducing flexibility

**After Fixes:**
- ✅ All constants defined at module level with clear names
- ✅ Detailed comments explaining IBMCEPH-15754 workaround
- ✅ Comprehensive inline documentation for pool creation logic
- ✅ All values configurable via constants or config

---

## Why Not 10/10?

**Minor Areas for Future Enhancement:**

1. **Type Hints** (-0.5 points)
   - Some functions could benefit from type annotations
   - Would improve IDE support and catch type errors earlier

2. **Test Assertion Workaround** (-0.5 points)
   - Commented assertions due to IBMCEPH-15754
   - While well-documented, full equality checks would be ideal
   - Waiting on upstream bug fix

---

## Code Quality Metrics

- **Lines Added:** ~1,800+
- **Files Modified:** 3
- **Test Scenarios:** 10 total (2 new)
- **Helper Functions:** 1 new, well-documented
- **Constants Defined:** 7 clear, purposeful constants
- **Documentation Coverage:** Excellent (inline comments + docstrings)
- **Bug Fixes Included:** Yes (convergence validation regex)

---

## Commit Impact Assessment

### Positive Impact ✅
- Enables EC pool testing for OSD upgrades
- Validates max flag behavior comprehensively
- Improves code maintainability significantly
- Sets good example for future contributions
- Well-documented workarounds for known issues

### Risk Assessment 🛡️
- **Low Risk**: Changes are additive, not modifying existing functionality
- **Well-Tested**: Multiple scenarios with proper cleanup
- **Documented Workarounds**: Known issues clearly explained
- **Backward Compatible**: Doesn't break existing tests

---

## Final Verdict

This is an **excellent, production-ready commit** that demonstrates:
- Strong software engineering practices
- Comprehensive testing methodology
- Clear documentation standards
- Thoughtful code organization
- Proper handling of technical debt

The commit successfully adds significant functionality while maintaining high code quality standards. With all identified issues addressed, this represents exemplary work that other developers should emulate.

### Recommendation: **APPROVE** ✅

This commit is ready for merge and sets a high bar for code quality in the project.

**Comment by s-vipin** [comment]
## Commit Review: Rating **7/10**

### Summary
Commit adds erasure-coded pool support for OSD ok-to-upgrade functionality and updates max flag tests across 5 files (~1,563 insertions, ~286 deletions).

### Strengths ✅

1. **Comprehensive Test Coverage**: Added scenarios 7-10 covering rack/chassis/host/OSD failure domains with max=100 and max=2
2. **Error Handling**: Added retry logic with 3 attempts and 10s delays in [`osd_ok_to_upgrade_utils.py`](ceph/rados/osd_ok_to_upgrade_utils.py:143-155)
3. **Code Reusability**: New helpers [`sort_osds_by_increasing_pg_count()`](tests/rados/test_ok_to_upgrade.py:1706) and [`create_pool_with_failure_domain()`](tests/rados/test_ok_to_upgrade.py:1770)
4. **Documentation**: Good docstrings and inline comments
5. **Bug Workarounds**: Properly documented IBMCEPH-15754 with TODO comments

### Critical Issues 🔴

1. **Uninitialized Variables** (Line 143-144): `out, err = [], []` remain empty if all retries fail, causing potential issues downstream
2. **Incomplete Exception Handling** (Line 151): Generic `Exception` caught but not logged when `ignore_errors=False`
3. **Inconsistent Assertions**: IBMCEPH-15754 workarounds vary - some check length, others fully commented out
4. **Magic Numbers**: Hardcoded `sleep_time = 10` and `range(3)` should be constants
5. **Regex Breaking Change** (Line 1862): Changed from `_check_offlines_pgs` to `operator()` without explanation
6. **Import Organization** (Line 17): `import time` added at end instead of with stdlib imports

### Recommendations 📋

- Fix retry logic to properly handle failures and log errors
- Add `MAX_RETRIES = 3` and `RETRY_SLEEP_TIME = 10` constants
- Standardize IBMCEPH-15754 workaround assertions
- Document regex pattern change reason
- Reorganize imports alphabetically

### Positive Aspects 🌟

- Excellent test scenario structure with clear steps
- Proper cleanup in finally blocks
- Handles both replicated and EC pools elegantly
- Good separation of concerns

**Overall**: Solid feature implementation with comprehensive tests, but retry logic needs fixing before merge.

**Comment by s-vipin** [comment]
## Commit Review: Rating **9/10** ⭐

### Summary
This commit adds erasure-coded pool support for OSD ok-to-upgrade functionality with comprehensive test coverage and improved error handling. The changes span 5 files with ~1,579 insertions and ~284 deletions.

### Strengths ✅

1. **Excellent Error Handling** (Lines 148-176): 
   - Proper variable initialization (`out, err = None, None`)
   - Comprehensive exception logging with attempt tracking
   - Retry logic with configurable constants (`MAX_RETRY_ATTEMPTS`, `RETRY_SLEEP_SECONDS`)
   - Clear error messages when all retries fail

2. **Code Quality**:
   - Well-defined constants instead of magic numbers
   - Proper exception tracking with `last_exception`
   - Conditional sleep (only between retries, not after last attempt)
   - Informative logging at each step

3. **Comprehensive Test Coverage**: 
   - Added scenarios 7-10 covering all failure domains (rack, chassis, host, OSD)
   - Tests both max=100 and max=2 flag variations
   - Includes negative test scenarios

4. **Code Reusability**: 
   - Helper functions like `sort_osds_by_increasing_pg_count()` and `create_pool_with_failure_domain()`
   - Good separation of concerns

5. **Documentation**: 
   - Clear docstrings and inline comments
   - Properly documented known issues (IBMCEPH-15754) with TODO comments

### Minor Issues 🟡

1. **Inconsistent Assertions**: Some tests use workaround assertions for IBMCEPH-15754 (checking length only), while others are fully commented out. Consider standardizing with a helper function.

2. **Regex Pattern Change**: Changed from `_check_offlines_pgs` to `operator()` in log parsing - this breaking change could benefit from a comment explaining why.

3. **Import Organization**: `import time` is with other imports but could be grouped with standard library imports at the top for better organization.

### What Makes This a 9/10:

✅ **Fixed all critical issues** from the previous review:
- Proper variable initialization
- Complete exception handling with logging
- No magic numbers

✅ **Production-ready code**:
- Robust retry mechanism
- Clear error messages for debugging
- Handles edge cases properly

✅ **Excellent test coverage** with multiple scenarios

✅ **Well-documented** with clear commit message

### Recommendation:
**Approved for merge** - This is high-quality, production-ready code with excellent error handling and comprehensive testing. The minor issues are cosmetic and don't impact functionality.

**Comment by s-vipin** [comment]
Pass logs #1 :- Failed at last test as it was running over night SSH timed out 
http://magna002.ceph.redhat.com/cephci-jenkins/vipin-runs/vips_automation_20260624-110936/

**Comment by s-vipin** [comment]
Pass logs :- http://magna002.ceph.redhat.com/cephci-jenkins/vipin-runs/vips_automation_20260625-042601/

**Review by pdhiran — APPROVED** [review]
looks good

File: suites/tentacle/rados/tier-2_rados_test_osd_deployment_workflows.yaml [hunk 1]
```diff
@@ -114,15 +114,15 @@ tests:
...
-     name: Add backup host node5 ( host with osd-bak label) to test OSD deployment workflows
-     desc: Add backup host node5 to test OSD deployment workflows
-     module: test_hosts.py
-     config:
-       deploy_osd: false
-       scenario: add_host
-       nodes:
-         - node5
-     abort-on-fail: true
+      name: Add backup host node5 ( host with osd-bak label) to test OSD deployment workflows
```
Comment (s-vipin) [inline suites/tentacle/rados/tier-2_rados_test_osd_deployment_workflows.yaml:117]: Space being added. Will update along with any review comments

**Review by harshkumarRH — APPROVED** [review]
Looks good
